types/module/server/module_key.go

```diff
@@ -1,20 +1,99 @@
 package server
 
 import (
+	"context"
+	"fmt"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc"
+
 	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/types/address"
 )
 
+// ModuleKey is an interface for module servers required by router.
 type ModuleKey interface {
 	types.InvokerConn
 
 	ModuleID() types.ModuleID
 	Address() sdk.AccAddress
+	Derive(key []byte) ModuleKey
+}
+
+type RootModuleKey interface {
+	ModuleKey
+	sdk.StoreKey
+}
+
+type moduleKey struct {
+	moduleName string
+	addr       []byte
+	i          InvokerFactory
+}
+
+// NewDerivedModuleKey creates a ModuleKey with a derived moduel address based on parent
+// module address and derivation key.
+func NewDerivedModuleKey(modName string, parentAddr, derivationKey []byte, i InvokerFactory) ModuleKey {
+	return moduleKey{modName, address.Derive(parentAddr, derivationKey), i}
+}
+
+// Invoker implements ModuleKey interface
+func (d moduleKey) Invoker(methodName string) (types.Invoker, error) {
+	return d.i(CallInfo{
+		Method: methodName,
+		Caller: d.ModuleID(),
+	})
 }
 
-type InvokerFactory func(callInfo CallInfo) (types.Invoker, error)
+// Invoke implements ModuleKey interface
+func (d moduleKey) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, _ ...grpc.CallOption) error {
+	invoker, err := d.Invoker(method)
+	if err != nil {
+		return err
+	}
+
+	return invoker(ctx, args, reply)
+}
+
+// NewStream implements ModuleKey interface
+func (d moduleKey) NewStream(context.Context, *grpc.StreamDesc, string, ...grpc.CallOption) (grpc.ClientStream, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+// ModuleID implements ModuleKey interface
```

Comment:
```suggestion
// moduleKey implements ModuleKey interface
```

Reply:
The method name is `ModuleID`, so the comment should start with the method name.